numpy/__init__.pyi

```diff
@@ -3058,7 +3058,12 @@ class flexible(generic): ...  # type: ignore
 # depending on whether or not it's used as an opaque bytes sequence
 # or a structure
 class void(flexible):
-    def __init__(self, value: _IntLike_co | bytes, /) -> None: ...
+    @overload
+    def __init__(self, value: _IntLike_co | bytes, /, dtype : None = ...) -> None: ...
+    @overload
+    def __init__(self, value: object, /, dtype : None = ...) -> None: NoReturn
+    @overload
+    def __init__(self, value: object, /, dtype: DTypeLike) -> None: ...
```

Comment:
```suggestion
    @overload
    def __init__(self, value: _IntLike_co | bytes, *, dtype: None = ...) -> None: ...
    @overload
    def __init__(self, value: Any, *, dtype: _DTypeLikeVoid) -> None: ...
```
Note that `_DTypeLikeVoid` still has to be imported from the `numpy._typing` namespace. And if I'm not mistaken `dtype` is keyword-only here, correct?

@seberg the more specific `_DTypeLikeVoid` union (a superset of all acceptable void-based dtype-likes) should help in dealing with this `None`-case, making the `NoReturn` overload redundant.

Secondly as a minor nit: `object` is generally reserved for situations where truly any arbitrary object is accepted (like in the case of `repr` for example), while `Any` moreso describes "not *all* objects are accepted, but I don't have good way to make it more specific".

Reply:
Thanks, I had not realized we have that one!

I had dropped the keyword-only, since Marten is right, there is no particular reasons to reject it here.

Reply:
Ah right, so if it's just a normal keyword-or-positional argument than there's no need for either the `/` or `*`. Might want to update the `np.void` docstring as well though to reflect that.

Reply:
Ah, right, thanks!

Reply:
Ah, the test failed using `"O,O"` as a void DTypeLike, but I will just change it to a list and consider it a small issue or feature of `_DTypeLikeVoid`

Reply:
> Ah, the test failed using `"O,O"` as a void DTypeLike

That's a good point actually, I forgot that `_DTypeLikeVoid` only contains specific string-literals in its definition.
Might be better to use `dtype: str | _DTypeLikeVoid` here rather than just plain `_DTypeLikeVoid`.

Reply:
OK, you are the typing boss :).